File: backend/src/lib/security.js
const crypto = require('crypto');

/**
 * Ký command bằng HMAC-SHA256 dùng API_KEY từ env.
 * Payload: id:agentId:action:params (params là chuỗi JSON từ DB).
 *
 * Agent sẽ verify chữ ký này trước khi thực thi lệnh.
 *
 * @param {object} command  - Prisma Command object
 * @returns {string}        - HMAC hex string
 */
function signCommand(command) {
  const secret = process.env.API_KEY;
  if (!secret) throw new Error('API_KEY chưa được cấu hình trong .env');

  const payload = `${command.id}:${command.agentId}:${command.action}:${command.params}`;
  return crypto.createHmac('sha256', secret).update(payload).digest('hex');
}

module.exports = { signCommand };

File: agent/modules/security.py
import hmac
import hashlib


class SecurityError(Exception):
    pass


def verify_and_extract_command(cmd: dict, api_secret: str) -> dict:
    """
    Verify HMAC-SHA256 signature của command trước khi thực thi.

    Payload phải khớp với backend/src/lib/security.js:
      id:agentId:action:params

    Raises SecurityError nếu signature vắng mặt hoặc không hợp lệ.
    Returns command dict (không có trường signature) nếu hợp lệ.
    """
    if not api_secret:
        raise SecurityError("api_key chưa được cấu hình — không thể verify")

    signature = cmd.get('signature')
    if not signature:
        raise SecurityError("Thiếu trường 'signature' trong command")

    params = cmd.get('params', '{}')
    if not isinstance(params, str):
        import json
        params = json.dumps(params)

    payload = f"{cmd.get('id')}:{cmd.get('agentId')}:{cmd.get('action')}:{params}"
    expected = hmac.new(api_secret.encode(), payload.encode(), hashlib.sha256).hexdigest()

    if not hmac.compare_digest(signature, expected):
        raise SecurityError("Signature không hợp lệ")

    return cmd
